File: frontend/src/api/publish_plan.ts
import request from './index'

export interface PublishPlan {
  id: number
  name: string
  target_upstream: number
  status: string
  total_count: number
  completed_count: number
  created_at: string
}

export interface PublishPlanItem {
  id: number
  plan_id: number
  package_id: number
  order: number
  category: string
  status: string
  selected: boolean
  error: string
  started_at: string
  completed_at: string
}

export interface PublishPlanDetail {
  plan: PublishPlan
  items: PublishPlanItem[]
}

export interface PublishPlansResponse {
  data: PublishPlan[]
  total: number
  page: number
  pageSize: number
}

export const getPublishPlans = (params?: { page?: number; pageSize?: number }) => {
  return request<PublishPlansResponse>({
    url: '/admin/publish-plans',
    method: 'get',
    params,
  })
}

export const getPublishPlan = (id: number) => {
  return request<PublishPlanDetail>({
    url: `/admin/publish-plans/${id}`,
    method: 'get',
  })
}

export const createPublishPlan = (data: {
  name: string
  target_upstream: number
  package_ids: number[]
  poll_interval?: number
  poll_timeout?: number
}) => {
  return request<PublishPlan>({
    url: '/admin/publish-plans',
    method: 'post',
    data,
  })
}

export const deletePublishPlan = (id: number) => {
  return request({
    url: `/admin/publish-plans/${id}`,
    method: 'delete',
  })
}

export const startPublishPlan = (id: number) => {
  return request({
    url: `/admin/publish-plans/${id}/start`,
    method: 'post',
  })
}

export const pausePublishPlan = (id: number) => {
  return request({
    url: `/admin/publish-plans/${id}/pause`,
    method: 'post',
  })
}

export const resumePublishPlan = (id: number) => {
  return request({
    url: `/admin/publish-plans/${id}/resume`,
    method: 'post',
  })
}

export const analyzePackages = (data: { package_ids: number[]; target_upstream: number }) => {
  return request<{
    packages: Array<{
      package_id: number
      organization: string
      name: string
      version: string
      sha256: string
      category: string
      selected: boolean
    }>
    publish_order: number[]
    total: number
  }>({
    url: '/admin/publish-plans/analyze',
    method: 'post',
    data,
  })
}

File: internal/models/publish_plan.go
package models

import "time"

// PublishPlan 发布计划
type PublishPlan struct {
	ID             int64     `xorm:"pk autoincr 'i_d'" json:"id"`
	Name           string    `xorm:"VARCHAR(200) NOT NULL 'name'" json:"name"`
	TargetUpstream int64     `xorm:"NOT NULL 'target_upstream'" json:"target_upstream"`
	Status         string    `xorm:"VARCHAR(20) NOT NULL 'status'" json:"status"`
	TotalCount     int       `xorm:"NOT NULL DEFAULT 0 'total_count'" json:"total_count"`
	CompletedCount int       `xorm:"NOT NULL DEFAULT 0 'completed_count'" json:"completed_count"`
	PollInterval   int       `xorm:"NOT NULL DEFAULT 60 'poll_interval'" json:"poll_interval"`
	PollTimeout    int       `xorm:"NOT NULL DEFAULT 600 'poll_timeout'" json:"poll_timeout"`
	CreatedAt      time.Time `xorm:"created 'created_at'" json:"created_at"`
	UpdatedAt      time.Time `xorm:"updated 'updated_at'" json:"updated_at"`
}

// PublishPlanItem 计划项
type PublishPlanItem struct {
	ID          int64     `xorm:"pk autoincr 'i_d'" json:"id"`
	PlanID      int64     `xorm:"NOT NULL INDEX 'plan_i_d'" json:"plan_id"`
	PackageID   int64     `xorm:"NOT NULL INDEX 'package_i_d'" json:"package_id"`
	Order       int       `xorm:"NOT NULL 'order'" json:"order"`
	Category    string    `xorm:"VARCHAR(20) NOT NULL 'category'" json:"category"`
	Status      string    `xorm:"VARCHAR(20) NOT NULL 'status'" json:"status"`
	Selected    bool      `xorm:"DEFAULT false 'selected'" json:"selected"`
	Error       string    `xorm:"TEXT 'error'" json:"error"`
	StartedAt   time.Time `json:"started_at"`
	CompletedAt time.Time `json:"completed_at"`
}
